### DISCUSSION TITLE
Frontend Pagination Issue

### DISCUSSION THREAD

**Discussion by bobwurtz** [livewire/livewire#3920] [1. Help] [answered]
Has anyone ever experienced an issue with Livewire pagination on the front end? The issue I am having is that sometimes Livewire registers a click on the wrong pagination element. This page is designed to show a list of menu items and then expand a specific item when the user clicks on it.

```
// ShowMenu.php (Livewire component)

public function render()
{
    $items = Item::whereIn('category_id', $this->categoryIds)
                                    ->where([
                                        ['status', true],
                                        ['deleted', false],
                                        ['price', '!=', null],
                                    ])
                                    ->search('items.name', $this->itemSearch)
                                    ->orderBy($this->sortField, $this->sortDirection)
                                    ->paginate(10);

    return view('livewire.customer.cafe.action.show-menu', compact('items'));
}


// livewire.customer.cafe.action.show-menu.blade.php

...
<!-- menu items -->
                    <div class="mt-2 overflow-y-auto space-y-2">

                        @foreach($items as $item)

                        <div class="rounded-lg shadow-xl overflow-hidden border-2 overflow-y-auto h-full p-2 cursor-pointer" x-on:click="itemDetail = true" wire:click="openItemDetail({{ $item->id }})">

                            <div class="grid grid-cols-4 gap-4">

                                <div class="col-span-3 flex flex-nowrap">

                                    {{ $item->name }}

                                </div>

                                <div class="col-span-1 text-right">

                                    ${{ number_format($item->price, 2, '.', ',') }}

                                </div>

                            </div>

                        </div>

                        @endforeach

                        {{ $items->links() }}

                    </div>
...
```

The `wire:click="openItemDetail({{ $item->id }})"` is what is giving me an issue. It works correctly most of the time, but sometimes the first element in the pagination list opens the wrong item. The only noticeable difference that I can find is that when the wrong item is opened, the `wire:click` event is at the beginning of the element in the rendered HTML (see screenshot below). For all elements except for the first one, the `wire:click` is at the end of the element like the code calls for. When I refresh the page and get a version of the page that works properly (no code changes, just a refresh), the `wire:click` event is at the end of the HTML element just like all the other items on the menu (second screenshot). I don't know enough about how browsers render code to know if this is important or not, but it was the only difference I could notice so I figured I would include it.
<br>

![Screen Shot 2021-10-01 at 9 53 30 AM](https://user-images.githubusercontent.com/59666225/135634033-56234f28-9f56-42dc-982d-4638e1976726.png)

![Screen Shot 2021-10-01 at 9 55 58 AM](https://user-images.githubusercontent.com/59666225/135634584-a19c6345-9ddc-441b-bd86-85bb30a0cb10.png)

<br>

Here is a video of the unexpected behavior:
<br>

https://user-images.githubusercontent.com/59666225/135635064-eb68bc3a-b9de-43ec-b8e2-efaef07bfcf8.mov
<br>

If any one has any ideas or suggestions I would appreciate it. Thank you!

**Comment by joshhanley** [comment, marked as answer]
@bobwurtz that's a morphdom issue, it's not diffing the elements correctly.

The reason is, Livewire has no way on its own to identify one row from the next. So instead you need to add a `wire:key` attribute whenever you have things in a loop.

Have a look at "dom diffing issues" section here https://laravel-livewire.com/docs/2.x/troubleshooting

Hope this helps!

**Reply by bobwurtz** [reply]
Looks like this fixed it. Thank you!

**Reply by joshhanley** [reply]
No worries! 🙂
